node/src/components/storage.rs

```diff
@@ -1,805 +1,442 @@
+            let block: Block = deser(raw_val);
+            // We use the opportunity for a small integrity check.
+            assert_eq!(
+                raw_key,
+                block.hash().as_ref(),
+                "found corrupt block in database"
+            );
+            let header = block.header();
+            if let Some(duplicate) = block_height_index.insert(header.height(), *block.hash()) {
+                warn!(
+                    height = %header.height(),
+                    hash_a = %header.hash(),
+                    hash_b = %duplicate,
+                    "found duplicate height in block database",
+                );
+            }
         }
-        .ignore()
+        info!("block store reindexing complete");
+        drop(cursor);
+        drop(block_tx);
+
+        Ok(Storage {
+            root,
+            env,
+            block_db,
+            deploy_db,
+            deploy_metadata_db,
+            block_height_index,
+            chainspec_cache,
+        })
     }
 
-    fn put_execution_results(
-        &self,
-        block_hash: <Self::Block as Value>::Id,
-        execution_results: HashMap<<Self::Deploy as Value>::Id, ExecutionResult>,
-        responder: Responder<()>,
-    ) -> Effects<Event<Self>>
+    /// Handles a storage request.
+    fn handle_storage_request<REv>(
+        &mut self,
+        req: StorageRequest,
+    ) -> Effects<<Self as Component<REv>>::Event>
     where
-        Self: Sized,
+        Self: Component<REv>,
     {
-        let deploy_store = self.deploy_store();
-        async move {
-            task::spawn_blocking(move || {
-                for (deploy_hash, execution_result) in execution_results.into_iter() {
-                    match deploy_store.put_execution_result(
-                        deploy_hash,
-                        block_hash,
-                        execution_result,
-                    ) {
-                        Ok(true) => (),
-                        Ok(false) => {
-                            warn!(%deploy_hash, %block_hash, "already stored execution result")
-                        }
-                        Err(error) => panic!(
-                            "failed to put execution results {} {}: {}",
-                            deploy_hash, block_hash, error
-                        ),
+        // Note: Database IO is handled in a blocking fashion on purpose throughout this function.
+        // The rationale behind is that long IO operations are very rare and cache misses
+        // frequent, so on average the actual execution time will be very low.
+        match req {
+            StorageRequest::PutBlock { block, responder } => {
+                let mut tx = self.env.rw_transaction();
+                let outcome = tx.put_value(self.block_db, block.hash(), &block);
+                tx.commit_ok();
+
+                if outcome {
+                    if let Some(prev) = self
+                        .block_height_index
```

Comment:
There should be a `block_by_height` LMDB-backed database and index cache should be capped by size to the last `n` elements. Inserts to `block_db` and `block_height_db` should be done in a single txn.

Reply:
I'm not sure we need to cap it. Assuming we hit 1 million blocks, 32 byte hash, 8 byte height and know about all of them, that's 10e6*(32+8) bytes 40 megs total. Even at 10x I think it's still reasonable (though at that point we'd be forced to commit the index to disk).

Reply:
I added a comment why it is an in-memory index currently. I'll happily volunteer to change that, should it become an issue due to reindexing or for other reasons. =)